kopykat-gradle-plugin/src/main/kotlin/fp/serrano/kopykat/KopyKatPlugin.kt

```diff
@@ -0,0 +1,97 @@
+package fp.serrano.kopykat
+
+import com.google.devtools.ksp.gradle.KspExtension
+import com.google.devtools.ksp.gradle.KspGradleSubplugin
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.kotlin.dsl.configure
+import org.gradle.kotlin.dsl.create
+import org.gradle.kotlin.dsl.withType
+import org.gradle.plugins.ide.idea.IdeaPlugin
+import org.gradle.plugins.ide.idea.model.IdeaModel
+import org.jetbrains.kotlin.gradle.dsl.kotlinExtension
+import org.jetbrains.kotlin.gradle.plugin.KotlinBasePlugin
+
+abstract class KopyKatPlugin : Plugin<Project> {
+
+    override fun apply(target: Project) {
+
+        val kkSettings = createKopyKatSettings(target)
+
+        configureKsp(target, kkSettings)
+
+        target.afterEvaluate {
+            // normally afterEvaluate should be avoided, but it's necessary here to allow KopyKatSettings to be
+            // configured otherwise applyKspPlugin will always be the default value.
+            if (kkSettings.applyKspPlugin.getOrElse(false)) {
+                target.pluginManager.apply("com.google.devtools.ksp")
+            }
+        }
+    }
+
+    private fun createKopyKatSettings(target: Project): KopyKatSettings {
+        return target.extensions.create<KopyKatSettings>(KOPYKAT_EXTENSION_NAME).apply {
+            applyKspPlugin.convention(true)
+
+            mutableCopy.convention(true)
+            copyMap.convention(true)
+            hierarchyCopy.convention(true)
+        }
+    }
+
+    private fun configureKsp(target: Project, kkSettings: KopyKatSettings) {
+        target.plugins.withType<KspGradleSubplugin> {
+
+            target.configureSourceSets()
+
+            target.configurations.matching { it.name == KspGradleSubplugin.KSP_MAIN_CONFIGURATION_NAME }.configureEach {
+                defaultDependencies {
+                    add(target.dependencies.create("com.github.kopykat-kt.kopykat:ksp:1.0-rc1"))
+                }
+            }
+
+            target.extensions.configure<KspExtension> {
+                arg("mutableCopy", "${kkSettings.mutableCopy.getOrElse(false)}")
+                arg("copyMap", "${kkSettings.copyMap.getOrElse(false)}")
+                arg("hierarchyCopy", "${kkSettings.hierarchyCopy.getOrElse(false)}")
+            }
```

Comment:
these don't seem to work. If I set `copyMap.set(false)` in Gradle, I can still use `person.copyMap {}`. I'm not sure if this is a Gradle or KSP thing. I'll investigate.

Reply:
Note that if `kkSettings` sets `mutableCopy` (or any other) to `null`, then the corresponding argument in KSP should **not** be set. And in any case, it should be set to `true` if nothing is provided.

Reply:
Good point - I've changed it to only set the KSP arg if the value is not-null.